dashboard/src/reducers/availablepackages.ts

```diff
@@ -67,13 +68,19 @@ const packageReducer = (
     case getType(actions.availablepackages.requestSelectedAvailablePackageVersions):
       return { ...state, isFetching: true };
     case getType(actions.availablepackages.receiveAvailablePackageSummaries): {
-      const isLastPage =
-        action.payload.page >= parseInt(action.payload.response.nextPageToken) ||
-        action.payload.response.nextPageToken === "";
+      // Note that the same condition identifies *before* we've fetched
+      // the first page and *after* we've fetched the last page. In both
+      // cases, the nextPageToken is empty. Since we have just fetched a
+      // page, only one option is possible.
+      const isLastPage = action.payload.response.nextPageToken === "";
       return {
         ...state,
         isFetching: false,
+        // TODO(minelson): Something is triggering a new request even after
+        // hasFinishedFetching is true, the response for which resets it.
+        // Or it could be that double requests are sent (seen in server logs)
```

Comment:
Haven't checked locally, but this is tricky... just a minimum changing altering this pagination state machine starts triggering unnecessary requests. It is a pain, but let's resolve it in a second iteration once we have everything working, so +1 as is.

Reply:
Oh - I actually meant to remove this comment as I solved that issue by adding the extra guard in the effect of Catalog.tsx which triggered the request:
```
    if (hasFinishedFetching) {
      return;
    }
```

The issue was that if the last page token changed, it would always initiate a new request, even if it'd finished fetching the whole series. So when it got to the last page, and the page token came back empty, it caused the extra request. Let me delete that comment.